### ISSUE TITLE
Database Structure

### ISSUE THREAD

**Issue by mdgilene** [rsanchez-wsu/jfiles#291]
The goal of this issue is to decide on the structure of the database that is going to be used to store users and the permissions to files/folders.

I have come up with a start and am looking for input and discussion on what might need changed.

```
+----------+
| Users    |
+----------+
| id       |  int
| password |  varchar
| name     |  varchar
| role     |  varchar
+----------+

+-----------+
| Roles     |
+-----------+
| id        | int
| name      | varchar
+-----------+

+--------------+
| Permissions  |
+--------------+
| id           | int
| doc          | xml
+--------------+

+-------------------+
| Role_Permissions  |
+-------------------+
| id                | int
| role_id           | int
| permission_id     | int
+-------------------+

+-----------------+
|User_Permissions |
+-----------------+
| id              | int
| user_id         | int
| permission_id   | int
+-----------------+
```

## Users
This table will be used to store the list of users and include things such as their name, ID, password and what permission group they are in (role).

## Roles
This table will be used for storing the list of possible roles a user could have each role would have a ID and a name.

## Permissions
This table would be responsible for storing the list of permissions for the file system. Each entry would have a id that would be used to represent that permission and it would have a block of XML containing the files/folders that this permissions grants access to.

## Role_Permissions
This table would assign permission ids to a role id. All users with this role would have access to the permissions granted to this role.

## User_Permissions
This table assigns permission ids to specific user ids. This is meant to be used for files/folders that only a specific user should have access to rather than an entire role.

## XML
In terms of the format of the XML that each permission would contain I was thinking something like the following would work. Each permission would have a fully valid and self contained XML document.
```
<?xml version="1.0" encoding="UTF-8"?>
<permission type="read">
	<directory path=""/>
	<directory path=""/>
	<file path="" />
	<file path="" />
</permission>
```
This permission would basically define 2 directories and 2 files to have read only access. The valid permission types would be `read`, `write`, and `readwrite`.

**Comment by rsanchez-wsu** [comment]
@mdgilene This is critically important and will impact nearly every part of the project.  Please make sure to mention it to everyone in Slack so that someone from every team sees it and has the opportunity to comment on it.

**Comment by mdgilene** [comment]
If no one has any comments on this setup then I am going to proceed with implementing this.

One minor thing to add to the above is that whatever role is created first and inserted into the Roles table will be considered the default role. Most likely a 'NONE' role will be used.

**Comment by cerasaurus** [comment]
I think this is clear, simple and effiecient. 
I was going to ask what the difference was between role_id/user_id and id but I see you took it out of roles and users. Did you mean to take it out of the permissions as well?

**Comment by mdgilene** [comment]
@cerasaurus Initially I gave each table a basic `id` column that was supposed to act as a simple table index and wasn't used for anything else but then I realized it was kind of redundant to have a table index id and an id for the actual role/permission. So I thought it would be better to just use the index in the table as the id of the role/permission. In regards to `role_id` and `user_id` these columns are going to act as "foreign keys" (I think this is the correct term for them @rsanchez-wsu?) which basically means it is an index into another table. So for example an entry in `Role_Permissions` that has a `role_id` of say 1, means that this permission is for the role defined in the `Roles` table with id (or position) 1.
